File: integrations/openai-agents/cortadel_openai_agents/_items.py
"""Helpers for reading the OpenAI Agents SDK's conversation items.

``TResponseInputItem`` is an alias for ``openai.types.responses.ResponseInputItemParam``
(``agents/items.py``: ``TResponseInputItem = ResponseInputItemParam``) — a union of ``TypedDict``\\ s,
so at runtime every item is a plain ``dict``. The runner normalizes items to dicts before they
reach a session (``agents/run_internal/items.py``: ``ensure_input_item_format`` →
``_coerce_to_dict``), but this module still accepts pydantic models and arbitrary objects so a
caller who hands us un-normalized items does not crash the agent.

Only *message* items carry text we can send to Cortadel. Function calls, function-call outputs,
reasoning items and the rest are agent plumbing: they are deliberately dropped rather than stored
as memories.
"""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any

__all__ = [
    "ROLES_TO_STORE",
    "item_as_mapping",
    "item_role",
    "item_text",
    "last_user_index",
    "last_user_text",
    "message_pairs",
]

#: Roles whose text is worth persisting as long-term memory. System/developer items are
#: instructions we (or the framework) authored, not things the user or agent said.
ROLES_TO_STORE = ("user", "assistant")

#: Content-part ``type`` values that carry plain text. ``input_text`` is what the user sends,
#: ``output_text`` is what the model produces, ``summary_text`` shows up inside reasoning items,
#: and bare ``text`` appears in some provider shims.
_TEXT_PART_TYPES = ("input_text", "output_text", "text", "summary_text")


def item_as_mapping(item: Any) -> Mapping[str, Any] | None:
    """Return ``item`` as a read-only mapping, or ``None`` if it is not dict-shaped.

    Handles the three shapes an item can realistically arrive in: a ``dict`` (the normal case),
    a pydantic model (``model_dump``), and an object exposing ``__dict__``.
    """
    if isinstance(item, Mapping):
        return item

    dump = getattr(item, "model_dump", None)
    if callable(dump):
        try:
            dumped = dump()
        except Exception:  # pragma: no cover - defensive; a broken model must not kill the run
            return None
        return dumped if isinstance(dumped, Mapping) else None

    raw = getattr(item, "__dict__", None)
    return raw if isinstance(raw, Mapping) else None


def item_role(item: Any) -> str | None:
    """Return the role of a *message* item, or ``None`` for anything else.

    An item counts as a message when it has a ``role`` and either omits ``type`` entirely (the
    shorthand ``{"role": "user", "content": "hi"}``) or sets it to ``"message"`` (the fully
    expanded shape the Responses API returns).
    """
    mapping = item_as_mapping(item)
    if mapping is None:
        return None

    item_type = mapping.get("type")
    if item_type is not None and item_type != "message":
        return None

    role = mapping.get("role")
    return role if isinstance(role, str) else None


def item_text(item: Any) -> str:
    """Extract the plain text of an item, flattening multi-part content.

    Returns an empty string when the item carries no text (an image-only turn, a refusal, a
    tool call). Callers treat empty as "nothing to remember".
    """
    mapping = item_as_mapping(item)
    if mapping is None:
        return ""

    content = mapping.get("content")
    if isinstance(content, str):
        return content.strip()

    if not isinstance(content, (list, tuple)):
        return ""

    chunks: list[str] = []
    for part in content:
        part_mapping = item_as_mapping(part)
        if part_mapping is None:
            if isinstance(part, str):
                chunks.append(part)
            continue
        part_type = part_mapping.get("type")
        if part_type is not None and part_type not in _TEXT_PART_TYPES:
            continue
        text = part_mapping.get("text")
        if isinstance(text, str) and text:
            chunks.append(text)

    return "\n".join(chunks).strip()


def message_pairs(items: Any) -> list[tuple[str, str]]:
    """Return ``(role, text)`` for every storable message item, in order.

    Items that are not messages, carry an unstorable role, or have no text are dropped.
    """
    pairs: list[tuple[str, str]] = []
    for item in items or ():
        role = item_role(item)
        if role not in ROLES_TO_STORE:
            continue
        text = item_text(item)
        if text:
            pairs.append((role, text))
    return pairs


def last_user_index(items: Any) -> int | None:
    """Return the index of the most recent user message that carries text."""
    sequence = list(items or ())
    for index in range(len(sequence) - 1, -1, -1):
        if item_role(sequence[index]) == "user" and item_text(sequence[index]):
            return index
    return None


def last_user_text(items: Any) -> str | None:
    """Return the text of the most recent user message, or ``None`` if there is none.

    This is the retrieval query for the automatic-memory filter: it is the question the model is
    about to answer, which is exactly what we want to search Cortadel for.
    """
    sequence = list(items or ())
    index = last_user_index(sequence)
    return None if index is None else item_text(sequence[index])

File: integrations/openai-agents/src/items.ts
/**
 * Helpers for reading the Agents SDK's conversation items.
 *
 * `AgentInputItem` (`@openai/agents-core/dist/types/aliases.d.ts`) is a discriminated union of
 * Zod-inferred plain objects — message items, function calls, function-call outputs, reasoning
 * items, compaction markers and more. At runtime every one of them is an ordinary object, so this
 * module reads them structurally and never narrows on a class.
 *
 * Only *message* items carry text we can send to Cortadel. Function calls, their outputs,
 * reasoning items and the rest are agent plumbing: they are deliberately dropped rather than
 * stored as memories.
 */

import type { AgentInputItem } from '@openai/agents';

/**
 * Roles whose text is worth persisting as long-term memory. System items are instructions we (or
 * the framework) authored, not things the user or the agent said.
 */
export const ROLES_TO_STORE = ['user', 'assistant'] as const;

export type StorableRole = (typeof ROLES_TO_STORE)[number];

/** One storable message, flattened to plain text. */
export interface MessagePair {
  role: StorableRole;
  text: string;
}

/**
 * Content-part `type` values that carry plain text. `input_text` is what the user sends,
 * `output_text` is what the model produces, `summary_text` shows up inside reasoning items, and
 * bare `text` appears in some provider shims.
 */
const TEXT_PART_TYPES = new Set(['input_text', 'output_text', 'text', 'summary_text']);

/**
 * Whether a content part's `type` names one of the text-bearing kinds.
 *
 * A missing `type` counts: some provider shims emit a bare `{ text: '…' }` part.
 *
 * The `typeof` check is load-bearing, not defensive noise. `type` arrives as `unknown`, and
 * coercing it with `String()` would flatten every object to `"[object Object]"` — worse, a part
 * whose `type` carried its own `toString()` could return `"text"` and impersonate a real text
 * part. Only an actual string can name a kind; anything else is a shape we do not understand and
 * is dropped.
 */
function isTextPartType(partType: unknown): boolean {
  if (partType === undefined) {
    return true;
  }
  return typeof partType === 'string' && TEXT_PART_TYPES.has(partType);
}

function asRecord(value: unknown): Record<string, unknown> | undefined {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : undefined;
}

/**
 * Return the role of a *message* item, or `undefined` for anything else.
 *
 * An item counts as a message when it has a `role` and either omits `type` entirely (the
 * shorthand `{ role: 'user', content: 'hi' }`) or sets it to `"message"` (the fully expanded
 * shape the Responses API returns). `type` is `z.ZodOptional<z.ZodLiteral<'message'>>` on all
 * three message schemas, so both forms are legal input.
 */
export function itemRole(item: unknown): string | undefined {
  const record = asRecord(item);
  if (!record) {
    return undefined;
  }
  const type = record['type'];
  if (type !== undefined && type !== 'message') {
    return undefined;
  }
  const role = record['role'];
  return typeof role === 'string' ? role : undefined;
}

/**
 * Extract the plain text of an item, flattening multi-part content.
 *
 * Returns an empty string when the item carries no text (an image-only turn, a refusal, a tool
 * call). Callers treat empty as "nothing to remember".
 */
export function itemText(item: unknown): string {
  const record = asRecord(item);
  if (!record) {
    return '';
  }

  const content = record['content'];
  if (typeof content === 'string') {
    return content.trim();
  }
  if (!Array.isArray(content)) {
    return '';
  }

  const chunks: string[] = [];
  for (const part of content) {
    if (typeof part === 'string') {
      chunks.push(part);
      continue;
    }
    const partRecord = asRecord(part);
    if (!partRecord) {
      continue;
    }
    if (!isTextPartType(partRecord['type'])) {
      continue;
    }
    const text = partRecord['text'];
    if (typeof text === 'string' && text) {
      chunks.push(text);
    }
  }

  return chunks.join('\n').trim();
}

function isStorableRole(role: string | undefined): role is StorableRole {
  return role === 'user' || role === 'assistant';
}

/**
 * Return `{ role, text }` for every storable message item, in order.
 *
 * Items that are not messages, carry an unstorable role, or have no text are dropped.
 */
export function messagePairs(items: readonly AgentInputItem[] | undefined): MessagePair[] {
  const pairs: MessagePair[] = [];
  for (const item of items ?? []) {
    const role = itemRole(item);
    if (!isStorableRole(role)) {
      continue;
    }
    const text = itemText(item);
    if (text) {
      pairs.push({ role, text });
    }
  }
  return pairs;
}

/** Index of the most recent user message that carries text, or `undefined`. */
export function lastUserIndex(items: readonly AgentInputItem[] | undefined): number | undefined {
  const sequence = items ?? [];
  for (let index = sequence.length - 1; index >= 0; index -= 1) {
    const item = sequence[index];
    if (itemRole(item) === 'user' && itemText(item)) {
      return index;
    }
  }
  return undefined;
}

/**
 * Text of the most recent user message, or `undefined` if there is none.
 *
 * This is the retrieval query for the automatic-memory filter: it is the question the model is
 * about to answer, which is exactly what we want to search Cortadel for.
 */
export function lastUserText(items: readonly AgentInputItem[] | undefined): string | undefined {
  const sequence = items ?? [];
  const index = lastUserIndex(sequence);
  return index === undefined ? undefined : itemText(sequence[index]);
}
